modules/mauth-authenticator-apachehttp/src/main/java/com/mdsol/mauth/apache/HttpClientPublicKeyProvider.java

```diff
@@ -96,18 +96,35 @@ private String getRequestUrlPath(UUID appUUID) {
   }
 
   private class PublicKeyResponseHandler implements ResponseHandler<String> {
+    private static final String MAX_AGE = "max-age";
+    private static final String PUBLIC_KEY_STR = "public_key_str";
 
     @Override
     public String handleResponse(HttpResponse response) throws IOException {
       if (response.getStatusLine().getStatusCode() == HttpStatus.SC_OK) {
+        if (ttl == null) {
+          ttl = getMaxAge(response).orElse(configuration.getTimeToLive());
+        }
+
         HttpEntity entity = response.getEntity();
         String responseAsString = EntityUtils.toString(entity, StandardCharsets.UTF_8);
         ObjectMapper mapper = new ObjectMapper();
-        return mapper.readTree(responseAsString).findValue("public_key_str").asText();
+        return mapper.readTree(responseAsString).findValue(PUBLIC_KEY_STR).asText();
       } else {
         throw new HttpClientPublicKeyProviderException("Invalid response code returned by server: "
-            + response.getStatusLine().getStatusCode());
+          + response.getStatusLine().getStatusCode());
+      }
+    }
+
+    public Optional<Long> getMaxAge(HttpResponse response) {
+      if (response.getFirstHeader(HttpHeaders.CACHE_CONTROL) != null) {
```

Comment:
I think this whole method can be simplified to:
```
      return Optional.ofNullable(response.getFirstHeader(HttpHeaders.CACHE_CONTROL))
              .map(header -> Long.parseLong(header.getValue()));
```

Reply:
Thanks for suggesting Optional.ofNullable in the first part! I included it in my commit to remove the if else statement, but aside from that,  I don't think the logic can be simplified further because I need to go through the cache-control header elements to get the max-age.

Reply:
Yes sorry I think I misunderstood the intention. 

Reading this a bit closer and looking at the HTTP header spec, I think it is valid to have the values split across multiple HTTP headers:

```
Cache-Control: public
Cache-Control: max-age=10000
```
(which is semantically equivalent to `Cache-Control: public, max-age=10000`)

In which case `getFirstHeader` will only give us `public`. Perhaps we should switch to using response.getHeaders(HttpHeaders.CACHE_CONTROL).flatMap(..) to handle this possibility?

Reply:
Oh wow I even didn't think that was even possible, so the exception is for headers with comma-separated values. I have resolved that scenario in my new commit.